### PR TITLE
Testing models build

### PR DESCRIPTION
**Pull request by joeljpoulin**
Adding tests gradually, testing low level components first.

### CODE REVIEW

**Comment by rdevon** [comment]
I think you might want to redo this PR to master?

**Comment by joeljpoulin** [comment]
@rdevon Yes, base branch has been changed to `master`.

File: cortex/_lib/parsing.py [hunk 1]
```diff
@@ -51,7 +51,7 @@ def parse_inputs(f):
 
 def parse_docstring(f):
     if f.__doc__ is None:
-        return {}
+        f.__doc__ = 'None\n None'
```
Comment (rdevon) [inline cortex/_lib/parsing.py]: Can we change these to "TODO"?

File: cortex/_lib/parsing.py [hunk 2]
```diff
@@ -63,6 +63,8 @@ def parse_docstring(f):
 
 
 def parse_header(f):
+    if f.__doc__ is None:
+        f.__doc__ = 'None\n None'
```
Comment (rdevon) [inline cortex/_lib/parsing.py]: Here as well

File: cortex/built_ins/models/classifier.py [hunk 3]
```diff
@@ -133,8 +133,7 @@ class ImageClassification(SimpleClassifier):
     defaults = dict(
         data=dict(batch_size=128, inputs=dict(inputs='images')),
         optimizer=dict(optimizer='Adam', learning_rate=1e-3),
-        train=dict(epochs=200, save_on_best='losses.classifier'),
-        classifier_args=dict(dropout=0.2))
+        train=dict(epochs=200, save_on_best='losses.classifier'))
```
Comment (rdevon) [inline cortex/built_ins/models/classifier.py]: What happened to `model=classifier_args=dict(dropout=0.2))`?

**Comment by joeljpoulin** [comment]
Changes following review made in commits faf9da7, b58f9a2 et 0abeb52.

File: tests/built_ins/networks/test_base_network.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
+from torch import nn
+
+
+def test_base_net(base_net_model):
+    """
+
+    Args:
+        base_net_model: BaseNet
+
+    Returns: True if BaseNet has an empty nn.Sequential models attribute
```
Comment (ReyhaneAskari) [inline tests/built_ins/networks/test_base_network.py]: It's not returning. You can say asserts. Same for the rest of the tests.

File: tests/test_argparsing.py [hunk 5]
```diff
@@ -46,16 +34,35 @@ def test_command_override_static():
...
 
 
-def test_static_override_parameters():
-    expected_type = 'resnet'
-    classifier_resnet = ClassifierModified()
+def test_static_override_parameters(args, classifier_modified):
+    """
+
+    Args:
+        args: Namespace
```
Comment (ReyhaneAskari) [inline tests/test_argparsing.py]: Please add a comment here and anywhere else that you are using the args from fixture.

Comment (joeljpoulin) [reply tests/test_argparsing.py]: Done in commit 41921c1.

File: tests/built_ins/networks/test_base_network.py (hunk 4 above)
Comment (joeljpoulin) [reply tests/built_ins/networks/test_base_network.py]: Done in commit a6efedc.

File: tests/built_ins/networks/test_convnets.py [hunk 6]
```diff
@@ -170,13 +186,15 @@ def test_simple_conv_encoder_forward(simple_conv_encoder_image_classification,
...
-    input_dim = simple_tensor_conv2d.dim()
+    # input_dim = simple_tensor_conv2d.dim()
     output = simple_conv_encoder_image_classification.forward(
         simple_tensor_conv2d)
-    output_dim = output.dim()
+    # output_dim = output.dim()
     equivalent = torch.equal(simple_tensor_conv2d, output)
-    assert input_dim == output_dim
+    # TODO: Test is failing on this assertion.
+    # assert input_dim == output_dim
```
Comment (joeljpoulin) [inline tests/built_ins/networks/test_convnets.py]: This might be an issue, as the dimension of the input is different of the dimension of the output. The test need to be verified if the error is in it or in Cortex code.

Comment (rdevon) [reply tests/built_ins/networks/test_convnets.py]: why would the input and output dim be the same for a conv encoder? The input is a 4d tensor (batch x channels x width x height) and the output is a 2d tensor (batch x length)

File: tests/built_ins/networks/test_convnets.py [hunk 7]
```diff
@@ -170,13 +186,15 @@ def test_simple_conv_encoder_forward(simple_conv_encoder_image_classification,
...
     assert equivalent is False
```
Comment (rdevon) [inline tests/built_ins/networks/test_convnets.py]: Why not just `assert not(equivalent)`?

File: tests/built_ins/networks/test_convnets.py [hunk 8]
```diff
@@ -162,6 +162,22 @@ def test_simple_net_init():
...
+             and that element-wise, the values have changed.
+
+    """
+    simple_net = SimpleNet()
+    input = torch.randn(128, 1, 32, 32)
+    output = simple_net.forward(input)
+    equivalent = torch.equal(input, output)
+    # TODO: Test is failing on this assertion.
+    # assert input.dim()== output.dim()
+    assert equivalent is False
```
Comment (rdevon) [inline tests/built_ins/networks/test_convnets.py]: `assert not equivalent`

Comment (joeljpoulin) [reply tests/built_ins/networks/test_convnets.py]: Changed in commit 07aee12 but change request does not seem outdated.

Comment (rdevon) [reply tests/built_ins/networks/test_convnets.py]: Ah you should update your pr from master using `merge`.

File: cortex/built_ins/networks/convnets.py [hunk 9]
```diff
@@ -45,8 +45,12 @@ def __init__(self, shape, dim_out=None, dim_h=64,
                  fully_connected_layers=None, nonlinearity='ReLU',
                  output_nonlinearity=None, f_size=4,
                  stride=2, pad=1, min_dim=4, n_steps=None,
+<<<<<<< HEAD
                  spectral_norm=False, **layer_args):
 
+=======
+                 spectral_norm=False, last_conv_nonlinearity=True, **layer_args):
+>>>>>>> master
```
Comment (rdevon) [inline cortex/built_ins/networks/convnets.py]: For instance, this isn't good.

File: tests/conftest.py [hunk 10]
```diff
@@ -174,16 +182,160 @@ def model_with_submodel(model_class_with_submodel, data_class):
...
+    from argparse import Namespace
+    import torch.nn as nn
+    return Namespace(
+        classifier_args={'dropout': 0.2},
+        classifier_type='convnet',
+        clean=False,
+        command=None,
+        config_file=None,
+        criterion=nn.CrossEntropyLoss(),
+        device=0,
```
Comment (rdevon) [inline tests/conftest.py]: Aren't these from args_mock? Is this file not being used?

This part needs to be changed to 'cpu':
https://github.com/rdevon/cortex/blob/master/tests/args_mock.py#L11

Otherwise it will fail on machines with GPU (in my tests.)
